src/bespokelabs/curator/request_processor/online_request_processor.py

```diff
@@ -0,0 +1,312 @@
+        status_tracker = StatusTracker()
+
+        # Get rate limits
+        rate_limits = self.get_rate_limits()
+        status_tracker.max_requests_per_minute = rate_limits["max_requests_per_minute"]
+        status_tracker.max_tokens_per_minute = rate_limits["max_tokens_per_minute"]
+        rpm = rate_limits["max_requests_per_minute"]
+
+        soft, hard = resource.getrlimit(resource.RLIMIT_NOFILE)
+        resource.setrlimit(
+            resource.RLIMIT_NOFILE,
+            (min(hard, int(10 * status_tracker.max_requests_per_minute)), hard),
+        )
+
+        # Track completed requests for resume functionality
+        completed_request_ids = set()
+        if os.path.exists(save_filepath) and resume:
+            with open(save_filepath, "r") as f:
+                for line in f:
+                    response = GenericResponse.model_validate_json(line)
+                    if not response.response_errors:
+                        completed_request_ids.add(response.generic_request.original_row_idx)
+
+        # Count total requests
+        total_requests = sum(1 for _ in open(generic_requests_filepath))
+
+        # Create progress bar
+        status_tracker.pbar = tqdm(
+            initial=len(completed_request_ids),
+            total=total_requests,
+            desc=f"Processing {self.__class__.__name__} requests",
+        )
+
+        # Use higher connector limit for better throughput
+        connector = aiohttp.TCPConnector(limit=10 * rpm)
+        async with aiohttp.ClientSession(
+            connector=connector
+        ) as session:  # Initialize ClientSession here
+            async with aiofiles.open(generic_requests_filepath) as file:
+                pending_requests = []
+
+                async for line in file:
+                    generic_request = GenericRequest.model_validate_json(line)
+
+                    if resume and generic_request.original_row_idx in completed_request_ids:
+                        status_tracker.num_tasks_already_completed += 1
+                        continue
+
+                    request = APIRequest(
+                        task_id=status_tracker.num_tasks_started,
+                        generic_request=generic_request,
+                        api_specific_request=self.create_api_specific_request(generic_request),
+                        attempts_left=max_attempts,
+                        prompt_formatter=self.prompt_formatter,
+                    )
+
+                    token_estimate = self.estimate_total_tokens(request.generic_request.messages)
+
+                    # Wait for capacity if needed
+                    while not status_tracker.has_capacity(token_estimate):
+                        await asyncio.sleep(0.1)
+
+                    # Consume capacity before making request
+                    status_tracker.consume_capacity(token_estimate)
+
+                    task = asyncio.create_task(
```

Comment:
Doesn't this only run if you are waiting somewhere else? 
https://github.com/bespokelabsai/curator/blob/main/src/bespokelabs/curator/request_processor/openai_online_request_processor.py#L417-L432
original code has a short sleep for this reason?

Reply:
Discussed offline, `async for` does this job